### PR TITLE
[Streaming Indexing] Introduce bulk HTTP API streaming flavor

### PR DESCRIPTION
**Pull request by reta**
### Description
Introduce bulk HTTP API streaming flavor

### Related Issues
Closes https://github.com/opensearch-project/OpenSearch/issues/9070

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugins/transport-reactor-netty4/src/main/java/org/opensearch/http/reactor/netty4/ReactorNetty4StreamingRequestConsumer.java [hunk 1]
```diff
@@ -23,10 +24,12 @@
 class ReactorNetty4StreamingRequestConsumer<T extends HttpContent> implements Consumer<T>, Publisher<HttpContent> {
     private final ReactorNetty4StreamingResponseProducer sender;
     private final StreamingHttpChannel httpChannel;
+    private final ByteBufAllocator alloc;
```
Comment (andrross) [inline plugins/transport-reactor-netty4/src/main/java/org/opensearch/http/reactor/netty4/ReactorNetty4StreamingRequestConsumer.java:27]: Where is this new field being used?

File: server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java [hunk 2]
```diff
@@ -15,10 +15,15 @@
...
 
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+
 /**
  * Wraps the instance of the {@link XContentBuilder} into {@link HttpChunk}
  */
 public final class XContentHttpChunk implements HttpChunk {
+    private static final byte[] CHUNK_SEPARATOR = new byte[] { '\r', '\n' };
```
Comment (andrross) [inline server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java:26]: Probably not super important, but is there a reason to prefer `\r\n` over the single byte `\n`?

File: server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java [hunk 3]
```diff
@@ -42,6 +47,13 @@ private XContentHttpChunk(@Nullable final XContentBuilder builder) {
         if (builder == null /* no content */) {
             content = BytesArray.EMPTY;
         } else {
+            // Always finalize the output chunk with '\r\n' sequence
+            try (final OutputStream out = builder.getOutputStream()) {
+                builder.close();
+                out.write(CHUNK_SEPARATOR);
```
Comment (andrross) [inline server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java:53]: This seems like we're going around the XContentBuilder to write directly to an output stream. The javadoc actually says this is ["dangerous"](https://github.com/opensearch-project/OpenSearch/blob/bf4367877eab27dff05a74d683d14e820130172d/libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java#L241). Could we instead just do the following?

```java
return CompositeBytesReference.of(BytesReference.bytes(builder), new BytesArray(CHUNK_SEPARATOR));
```

File: server/src/main/java/org/opensearch/rest/action/document/RestBulkStreamingAction.java [hunk 4]
```diff
@@ -95,6 +96,16 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         final Boolean defaultRequireAlias = request.paramAsBoolean(DocWriteRequest.REQUIRE_ALIAS, null);
         final TimeValue timeout = request.paramAsTime("timeout", BulkShardRequest.DEFAULT_TIMEOUT);
         final String refresh = request.param("refresh");
+        final TimeValue batch_interval = request.paramAsTime("batch_interval", null);
+        final int batch_size = request.paramAsInt("batch_size", 1); /* by default, batch size of 1 */
```
Comment (andrross) [inline server/src/main/java/org/opensearch/rest/action/document/RestBulkStreamingAction.java:100]: The Java fields should be camelCase.

File: plugins/transport-reactor-netty4/src/main/java/org/opensearch/http/reactor/netty4/ReactorNetty4StreamingRequestConsumer.java (hunk 1 above)
Comment (reta) [reply plugins/transport-reactor-netty4/src/main/java/org/opensearch/http/reactor/netty4/ReactorNetty4StreamingRequestConsumer.java:27]: Not used anymore, removing, thanks!

File: server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java (hunk 2 above)
Comment (reta) [reply server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java:26]: Yes, for example, the (single) chunk in response may looks like this: 
```
{ "index": { "_index": "test-streaming", "_id": "3" } }
{ "name": "john" }
```

(we don't emit those but it is possible to have multi-line here)

File: server/src/main/java/org/opensearch/rest/action/document/RestBulkStreamingAction.java (hunk 4 above)
Comment (reta) [reply server/src/main/java/org/opensearch/rest/action/document/RestBulkStreamingAction.java:100]: Oh man, my bad!

File: server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java (hunk 3 above)
Comment (reta) [reply server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java:53]: > This seems like we're going around the XContentBuilder to write directly to an output stream.

Correct, this is why we close the stream before writing, but `CompositeBytesReference` looks less "hacky", let me try it, thanks!

File: server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java (hunk 2 above)
Comment (andrross) [reply server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java:23]: Ooooh, so lines within a chunk will be separated by `\n` and chunks are separated by `\r\n`?

Comment (andrross) [reply server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java:23]: If that's the case, my instinct is that given that newline delimited JSON is human readable and that `\n` and `\r\n` are rendered the same, maybe we should use a printable character as part of the chunk delimiter?

Comment (reta) [reply server/src/main/java/org/opensearch/common/xcontent/support/XContentHttpChunk.java:23]: > If that's the case, my instinct is that given that newline delimited JSON is human readable and that `\n` and `\r\n` are rendered the same, maybe we should use a printable character as part of the chunk delimiter?

Chunks are separated by `\r\n` (see please https://en.wikipedia.org/wiki/Chunked_transfer_encoding), it just we have to kind of keep the separator within chunk itself.

**Review by andrross — APPROVED** [review]

**Comment by reta** [comment]
@andrross thanks a mill for approving, conflicts resolved + more tests added, could you please reapprove? thank you!

**Review by andrross — APPROVED** [review]

File: client/rest/src/main/java/org/opensearch/client/RestClient.java [hunk 5]
```diff
@@ -431,6 +437,48 @@ private Publisher<Message<HttpResponse, Publisher<ByteBuffer>>> streamRequest(
         });
     }
 
+    /**
+     * Frame the {@link ByteBuffer} into individual chunks that are separated by '\r\n' sequence.
+     * @param b {@link ByteBuffer} to split
+     * @return individual chunks
+     */
+    private static Collection<ByteBuffer> frame(ByteBuffer b) {
```
Comment (andrross) [inline client/rest/src/main/java/org/opensearch/client/RestClient.java:445]: This looks like a good candidate for a unit test. What do you think?

Comment (reta) [reply client/rest/src/main/java/org/opensearch/client/RestClient.java:445]: My apologies, have seen the comment after merge only. Sadly our `codecov` captures only limited subset of tests, the `ReactorNetty4StreamingIT` covers this method in full:

image

**Pull request merged**